src/reconstruct.rs

```diff
@@ -0,0 +1,132 @@
+
+// Argument constants
+static INPUT: &str = "INPUT";
+static OUTPUT: &str = "OUTPUT";
+
+pub struct ReconstructApp {
+    input_filename: String,
+    input_format: QuantizedEmbeddingFormat,
+    output_filename: String,
+    n_threads: usize,
+}
+
+impl FinalfusionApp for ReconstructApp {
+    fn app() -> App<'static, 'static> {
+        App::new("reconstruct")
+            .about("Reconstruct embedding matrix from quantized embeddings")
+            .arg(
+                Arg::with_name(INPUT)
+                    .help("Finalfusion model")
+                    .index(1)
+                    .required(true),
+            )
+            .arg(
+                Arg::with_name(OUTPUT)
+                    .help("Output finalfusion model")
+                    .index(2)
+                    .required(true),
+            )
+            .arg(
+                Arg::with_name(N_THREADS)
+                    .short("t")
+                    .long("threads")
+                    .value_name("N")
+                    .help("Number of threads (default: logical_cpus /2)")
+                    .takes_value(true),
+            )
+            .arg(
+                Arg::with_name(INPUT_FORMAT)
+                    .short("f")
+                    .long("from")
+                    .value_name("FORMAT")
+                    .takes_value(true)
+                    .possible_values(&["finalfusion", "finalfusion_mmap"])
+                    .default_value("finalfusion"),
+            )
+    }
+
+    fn parse(matches: &ArgMatches) -> Self {
+        // Arguments
+        let input_filename = matches.value_of(INPUT).unwrap().to_owned();
+        let output_filename = matches.value_of(OUTPUT).unwrap().to_owned();
+
+        // Options
+        let n_threads = matches
+            .value_of(N_THREADS)
+            .map(|a| a.parse().or_exit("Cannot parse number of threads", 1))
+            .unwrap_or(num_cpus::get() / 2);
+
+        let input_format = matches
+            .value_of(INPUT_FORMAT)
+            .map(|v| QuantizedEmbeddingFormat::try_from(v).or_exit("Cannot parse input format", 1))
+            .unwrap();
+
+        ReconstructApp {
+            input_filename,
+            input_format,
+            output_filename,
+            n_threads,
+        }
+    }
+
+    fn run(&self) {
+        env_logger::init();
+
+        ThreadPoolBuilder::new()
+            .num_threads(self.n_threads)
+            .build_global()
+            .unwrap();
+        let embeddings = read_quantized_embeddings(&self.input_filename, self.input_format)
+            .or_exit("Cannot read embeddings", 1);
+        let len = embeddings.vocab().vocab_len();
+        let (metadata, vocab, storage, norms) = embeddings.into_parts();
+
+        // Reconstruct storage
+        let reconstructed_storage = reconstruct_storage(&storage);
+        let nd_norms = match norms {
+            None => NdNorms::new(Array1::ones(len)),
+            Some(inner) => inner,
+        };
```

Comment:
Change 8: construct an ```Array1``` and pass that to the constructor in case ```norms``` is ```None``` (coz before I just ```Embeddings::new(metadata, vocab, reconstructed_storage, norms.unwrap());```)